04-materials/03-debugging.md

```diff
@@ -1,111 +1,431 @@
+- `404 Not Found` (failed API request)
+- `NetworkError: Failed to fetch` (network connection issue)
+
+Runtime errors show up in the browser console or `jupyter lab` terminal logs after your extension is loaded.
+
+:::{important} Quick Reference
+:class: simple
+:icon: false
+
+| Error Type | Where to Look | When It Happens |
+|------------|---------------|-----------------|
+| **Build-time** | Terminal (`jlpm build`) | During compilation |
+| **Runtime (Frontend)** | Browser console | After extension loads |
+| **Runtime (Backend)** | `jupyter lab` terminal logs | After extension loads |
+:::
+
+#### 🗺️ By where they occur
+
+With a full stack extension, which includes a server component, it is important to understand which side, server-side or client-side, is causing the error.
+
+##### Python server errors
+
+Python errors will show up in the terminal from which you launched your JupyterLab instance (the server terminal).
+
+**Example:**
+```python
+ValueError: Invalid endpoint configuration
+  File "jupytercon2025_extension_workshop/routes.py", line 32, in get
+```
+
+
+##### Javascript frontend errors
+
+JavaScript error details will be available in the browser console.
+
+**Example:**
+```javascript
+Uncaught TypeError: widget.render is not a function
+    at activate (index.ts:23)
+```
+
+:::{tip}
+There may be times where an issue initially appears to be a frontend problem, for example, data not being displayed, but it is actually a backend problem, like an API endpoint returning the wrong data, or a 500 error.
+:::
+
+### 📚 Reading Stack Traces
 
-### Types of Errors
-- Build-time errors vs runtime errors
-- Python backend errors vs TypeScript frontend errors
+- **Identifying the source of the error**:
+  When JavaScript throws an error, it gives you a stack trace, a trail showing the path the code took before it crashed. It may seem overwhelming at first, but starting from the top and looking for recognizable file names can help. While the stack trace might show you multiple function calls usually only one or two of those are in _your_ code. The others are likely in other libraries like JupyterLab itself. Finding your line in the code, you can find out where things have gone wrong.
 
-### Reading Stack Traces
-- Minified vs source-mapped traces
-- JavaScript stack traces
-- Identifying the source of the error
+- **Minified vs source-mapped traces**:
+  Also important, if you see names like `bundle.js:4323`, that's minified code. In development mode with {term}`source maps <Source maps>` enabled, you'll see actual files names like `Panel.tsx:53`, a much more helpful identifier. Extensions created with the `extension-template` should have source maps enabled already.
 
-## 2. Terminal
+## 2️⃣ Terminal
+
+### 📋 Terminal Logs
 
-### Terminal Logs
 - Where to find JupyterLab server logs
+  In the terminal where you run `jupyter lab`, you'll find a great deal of information.
```

Comment:
```suggestion
  
  In the terminal where you run `jupyter lab`, you'll find a great deal of information.
```

I think you intended for this to show up on a separate line, correct? Currently it's rendering all as one sentence: "Where to find JupyterLab server logs In the terminal..."

Since this is a one-item bulleted list, maybe we can get rid of the bullet instead?